File: web/src/services/debrid/TorBoxService.ts
// Port of Sources/DebridStreamer/Services/Debrid/TorBoxService.swift.
//
// fetch-based TorBox client. Mirrors the Swift actor: /torrents/checkcached
// cache check, /torrents/createtorrent add, the /torrents/mylist snapshot +
// best-file selection + /torrents/requestdl stream-URL flow (with the
// "throw instead of streaming file_id=0 when not ready" guard), and /user/me
// account decoding. Auth is a Bearer header only (no apikey in query/body). The
// poll SLEEP is injectable (default no-op) so tests don't wait.

import {
  CacheStatus,
  type DebridAccountInfo,
  type DebridFileCandidate,
  DebridFileSelector,
  type DebridServiceType,
  DebridServiceType as DebridServiceTypeNS,
  type StreamInfo,
  AudioFormat,
  SourceType,
  VideoCodec,
  VideoQuality,
} from "./models";
import {
  type DebridService,
  DebridError,
  type FetchImpl,
  defaultFetchImpl,
} from "./types";
import {
  asObject,
  asObjectArray,
  int64Value,
  noopSleep,
  parseJSONObject,
  type Sleep,
} from "./jsonHelpers";

interface TorrentFileEntry {
  id: number;
  name: string;
  sizeBytes: number;
}

interface TorrentSnapshot {
  state: string;
  files: TorrentFileEntry[];
}

function chunked<T>(arr: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

export class TorBoxService implements DebridService {
  readonly serviceType: DebridServiceType = DebridServiceTypeNS.torBox;
  private readonly apiToken: string;
  private readonly baseURL = "https://api.torbox.app/v1/api";
  private readonly fetchImpl: FetchImpl;
  private readonly sleep: Sleep;

  constructor(apiToken: string, fetchImpl?: FetchImpl, sleep: Sleep = noopSleep) {
    this.apiToken = apiToken;
    this.fetchImpl = fetchImpl ?? defaultFetchImpl();
    this.sleep = sleep;
  }

  async checkCache(hashes: string[]): Promise<Record<string, CacheStatus>> {
    if (hashes.length === 0) return {};

    const results: Record<string, CacheStatus> = {};
    for (const chunk of chunked(hashes, 100)) {
      const hashParam = chunk.join(",");
      const data = await this.requestRaw(
        "/torrents/checkcached",
        "GET",
        `hash=${hashParam}&format=object`,
      );

      const json = parseJSONObject(data);
      const dataObj = json && asObject(json.data);
      if (dataObj) {
        for (const hash of chunk) {
          const lowerHash = hash.toLowerCase();
          if (Object.prototype.hasOwnProperty.call(dataObj, lowerHash)) {
            results[lowerHash] = CacheStatus.cached(null, null, null);
          } else {
            results[lowerHash] = CacheStatus.notCached;
          }
        }
      }
    }

    return results;
  }

  async addMagnet(hash: string): Promise<string> {
    const magnet = `magnet:?xt=urn:btih:${hash}`;
    const body = `magnet=${encodeURIComponent(magnet)}`;
    const data = await this.requestRaw("/torrents/createtorrent", "POST", undefined, body);

    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    const id = dataObj ? int64Value(dataObj.torrent_id) : null;
    // The Swift impl requires torrent_id to be an Int specifically.
    if (id == null || typeof dataObj?.torrent_id !== "number") {
      throw DebridError.downloadFailed("Failed to add magnet to TorBox");
    }
    return String(id);
  }

  async selectFiles(_torrentId: string, _fileIds: number[]): Promise<void> {
    // TorBox handles file selection during creation.
  }

  async getStreamURL(torrentId: string): Promise<StreamInfo> {
    const maxAttempts = 20;
    const terminalReadyStates = new Set(["cached", "completed", "uploading"]);
    let selectedFile: TorrentFileEntry | null = null;
    let lastState = "";

    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const snapshot = await this.getTorrentSnapshot(torrentId);
      lastState = snapshot.state.toLowerCase();
      const best = this.bestFile(snapshot.files);
      if (best != null) {
        selectedFile = best;
        break;
      }

      if (lastState.includes("stalled")) {
        throw DebridError.downloadFailed(`Torrent stalled: ${snapshot.state}`);
      }

      if (terminalReadyStates.has(lastState)) {
        break;
      }

      if (attempt < maxAttempts - 1) {
        await this.sleep(1000);
      }
    }

    if (selectedFile == null && !terminalReadyStates.has(lastState)) {
      throw DebridError.downloadFailed(
        `Torrent not ready: ${lastState.length === 0 ? "unknown" : lastState}`,
      );
    }

    const fallbackId = 0;
    const fileId = selectedFile?.id ?? fallbackId;
    const fileName = selectedFile?.name ?? "TorBox Stream";
    const size = selectedFile?.sizeBytes ?? 0;

    const data = await this.requestRaw(
      "/torrents/requestdl",
      "GET",
      `torrent_id=${torrentId}&file_id=${fileId}&zip_link=false`,
    );

    const json = parseJSONObject(data);
    const dataStr = json && typeof json.data === "string" ? json.data : null;
    if (dataStr == null) throw DebridError.noFilesAvailable();

    return {
      streamURL: dataStr,
      quality: VideoQuality.parse(fileName),
      codec: VideoCodec.parse(fileName),
      audio: AudioFormat.parse(fileName),
      source: SourceType.parse(fileName),
      sizeBytes: size,
      fileName,
      debridService: "TB",
    };
  }

  async unrestrict(link: string): Promise<string> {
    if (!isAbsoluteURL(link)) throw DebridError.downloadFailed("Invalid link");
    return link;
  }

  async validateToken(): Promise<boolean> {
    try {
      await this.getAccountInfo();
      return true;
    } catch {
      return false;
    }
  }

  async getAccountInfo(): Promise<DebridAccountInfo> {
    const data = await this.requestRaw("/user/me", "GET");
    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    if (dataObj == null) throw DebridError.invalidToken();

    const email = typeof dataObj.email === "string" ? dataObj.email : "Unknown";
    const premium = typeof dataObj.plan === "number" ? dataObj.plan : 0;

    let premiumExpiry: Date | null = null;
    if (typeof dataObj.premium_expires_at === "string") {
      const parsed = new Date(dataObj.premium_expires_at);
      premiumExpiry = Number.isNaN(parsed.getTime()) ? null : parsed;
    }

    return {
      username: email,
      email,
      premiumExpiry,
      isPremium: premium > 0,
    };
  }

  // MARK: - HTTP

  private async requestRaw(
    path: string,
    method: string,
    queryParams?: string,
    body?: string,
  ): Promise<string> {
    let urlStr = this.baseURL + path;
    if (queryParams != null) {
      urlStr += `?${queryParams}`;
    }

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiToken}`,
    };
    if (body != null) {
      headers["Content-Type"] = "application/x-www-form-urlencoded";
    }

    const response = await this.fetchImpl(urlStr, { method, headers, body });
    const status = response.status;

    if (!(status >= 200 && status <= 299)) {
      if (status === 401 || status === 403) throw DebridError.invalidToken();
      const errText = (await response.text().catch(() => "")) || "";
      throw DebridError.httpError(status, errText);
    }

    return response.text();
  }

  private async getTorrentSnapshot(torrentId: string): Promise<TorrentSnapshot> {
    const data = await this.requestRaw(
      "/torrents/mylist",
      "GET",
      `id=${torrentId}&bypass_cache=true`,
    );

    const json = parseJSONObject(data);
    if (json == null) throw DebridError.torrentNotFound(torrentId);

    const torrentObject = this.extractTorrentObject(json, torrentId);
    if (torrentObject == null) throw DebridError.torrentNotFound(torrentId);

    const state =
      typeof torrentObject.download_state === "string"
        ? torrentObject.download_state
        : "";
    const files = this.parseFiles(torrentObject.files);
    return { state, files };
  }

  private extractTorrentObject(
    json: Record<string, unknown>,
    torrentId: string,
  ): Record<string, unknown> | null {
    const object = asObject(json.data);
    if (object) return object;

    const list = asObjectArray(json.data);
    if (list == null) return null;

    const torrentID = int64Value(torrentId);
    if (torrentID != null) {
      const match = list.find((entry) => int64Value(entry.id) === torrentID);
      if (match) return match;
    }

    return list[0] ?? null;
  }

  private parseFiles(rawFiles: unknown): TorrentFileEntry[] {
    const files = asObjectArray(rawFiles);
    if (files == null) return [];

    const out: TorrentFileEntry[] = [];
    for (const file of files) {
      const fileId64 = int64Value(file.id);
      if (fileId64 == null) continue;
      const name =
        (typeof file.name === "string" ? file.name : null) ??
        (typeof file.short_name === "string" ? file.short_name : null) ??
        "Unknown";
      const size = int64Value(file.size) ?? 0;
      out.push({ id: fileId64, name, sizeBytes: size });
    }
    return out;
  }

  private bestFile(files: TorrentFileEntry[]): TorrentFileEntry | null {
    if (files.length === 0) return null;

    const candidates: DebridFileCandidate[] = files.map((file) => ({
      link: String(file.id),
      fileName: file.name,
      sizeBytes: file.sizeBytes,
    }));

    const selected = DebridFileSelector.selectBest(candidates);
    if (selected == null) return null;
    const selectedId = Number.parseInt(selected.link, 10);
    if (Number.isNaN(selectedId)) return null;

    return files.find((f) => f.id === selectedId) ?? null;
  }
}

function isAbsoluteURL(value: string): boolean {
  try {
    new URL(value);
    return true;
  } catch {
    return false;
  }
}

File: Sources/DebridStreamer/Services/Debrid/TorBoxService.swift
import Foundation

/// TorBox API integration.
/// API docs: https://torbox.app/api
actor TorBoxService: DebridServiceProtocol {
    let serviceType = DebridServiceType.torBox
    private let apiToken: String
    private let baseURL = "https://api.torbox.app/v1/api"
    private let session: URLSession

    private struct TorrentFileEntry {
        let id: Int
        let name: String
        let sizeBytes: Int64
    }

    private struct TorrentSnapshot {
        let state: String
        let files: [TorrentFileEntry]
    }

    // Reused across calls instead of allocating per `getAccountInfo`.
    private static let premiumExpiryFormatter = ISO8601DateFormatter()

    init(apiToken: String, session: URLSession = AppHTTP.api) {
        self.apiToken = apiToken
        self.session = session
    }

    func checkCache(hashes: [String]) async throws -> [String: CacheStatus] {
        guard !hashes.isEmpty else { return [:] }

        var results: [String: CacheStatus] = [:]
        let chunks = hashes.chunked(into: 100)

        for chunk in chunks {
            let hashParam = chunk.joined(separator: ",")
            let data = try await requestRaw(
                path: "/torrents/checkcached",
                method: "GET",
                queryParams: "hash=\(hashParam)&format=object"
            )

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let dataObj = json["data"] as? [String: Any] {
                for hash in chunk {
                    let lowerHash = hash.lowercased()
                    if let _ = dataObj[lowerHash] {
                        results[lowerHash] = .cached(fileId: nil, fileName: nil, fileSize: nil)
                    } else {
                        results[lowerHash] = .notCached
                    }
                }
            }
        }

        return results
    }

    func addMagnet(hash: String) async throws -> String {
        let magnet = "magnet:?xt=urn:btih:\(hash)"
        let body = "magnet=\(magnet.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? magnet)"
        let data = try await requestRaw(path: "/torrents/createtorrent", method: "POST", body: body)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any],
              let id = dataObj["torrent_id"] as? Int else {
            throw DebridError.downloadFailed("Failed to add magnet to TorBox")
        }
        return String(id)
    }

    func selectFiles(torrentId: String, fileIds: [Int]) async throws {
        // TorBox handles file selection during creation
    }

    func getStreamURL(torrentId: String) async throws -> StreamInfo {
        let maxAttempts = 20
        let terminalReadyStates: Set<String> = ["cached", "completed", "uploading"]
        var selectedFile: TorrentFileEntry?
        var lastState = ""

        for attempt in 0..<maxAttempts {
            let snapshot = try await getTorrentSnapshot(torrentId: torrentId)
            lastState = snapshot.state.lowercased()
            if let best = bestFile(from: snapshot.files) {
                selectedFile = best
                break
            }

            if lastState.contains("stalled") {
                throw DebridError.downloadFailed("Torrent stalled: \(snapshot.state)")
            }

            // If the torrent is already complete/cached but file metadata is missing,
            // skip waiting and fall back to TorBox default file selection.
            if terminalReadyStates.contains(lastState) {
                break
            }

            if attempt < maxAttempts - 1 {
                try await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }

        // Only fall back to file_id=0 when we have no concrete file AND the torrent
        // reached a terminal/ready state. If it's still processing (or in an unknown
        // non-terminal state) after exhausting attempts, surface a failure instead of
        // blindly streaming file 0.
        if selectedFile == nil && !terminalReadyStates.contains(lastState) {
            throw DebridError.downloadFailed("Torrent not ready: \(lastState.isEmpty ? "unknown" : lastState)")
        }

        let fallbackId = 0
        let fileId = selectedFile?.id ?? fallbackId
        let fileName = selectedFile?.name ?? "TorBox Stream"
        let size = selectedFile?.sizeBytes ?? 0

        let data = try await requestRaw(
            path: "/torrents/requestdl",
            method: "GET",
            queryParams: "torrent_id=\(torrentId)&file_id=\(fileId)&zip_link=false"
        )

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataStr = json["data"] as? String else {
            throw DebridError.noFilesAvailable
        }

        return StreamInfo(
            streamURL: dataStr,
            quality: VideoQuality.parse(from: fileName),
            codec: VideoCodec.parse(from: fileName),
            audio: AudioFormat.parse(from: fileName),
            source: SourceType.parse(from: fileName),
            sizeBytes: size,
            fileName: fileName,
            debridService: "TB"
        )
    }

    func unrestrict(link: String) async throws -> URL {
        guard let url = URL(string: link) else {
            throw DebridError.downloadFailed("Invalid link")
        }
        return url
    }

    func validateToken() async throws -> Bool {
        do {
            _ = try await getAccountInfo()
            return true
        } catch {
            return false
        }
    }

    func getAccountInfo() async throws -> DebridAccountInfo {
        let data = try await requestRaw(path: "/user/me", method: "GET")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any] else {
            throw DebridError.invalidToken
        }

        let email = dataObj["email"] as? String ?? "Unknown"
        let premium = dataObj["plan"] as? Int ?? 0

        var premiumExpiry: Date?
        if let expiryStr = dataObj["premium_expires_at"] as? String {
            premiumExpiry = Self.premiumExpiryFormatter.date(from: expiryStr)
        }

        return DebridAccountInfo(
            username: email,
            email: email,
            premiumExpiry: premiumExpiry,
            isPremium: premium > 0
        )
    }

    // MARK: - HTTP

    private func requestRaw(path: String, method: String, queryParams: String? = nil, body: String? = nil) async throws -> Data {
        var urlStr = baseURL + path
        if let queryParams = queryParams {
            urlStr += "?\(queryParams)"
        }

        guard let url = URL(string: urlStr) else {
            throw DebridError.networkError("Invalid URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(apiToken)", forHTTPHeaderField: "Authorization")

        if let body = body {
            request.httpBody = body.data(using: .utf8)
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DebridError.networkError("Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                throw DebridError.invalidToken
            }
            throw DebridError.httpError(httpResponse.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        return data
    }

    private func getTorrentSnapshot(torrentId: String) async throws -> TorrentSnapshot {
        let data = try await requestRaw(
            path: "/torrents/mylist",
            method: "GET",
            queryParams: "id=\(torrentId)&bypass_cache=true"
        )

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DebridError.torrentNotFound(torrentId)
        }

        guard let torrentObject = extractTorrentObject(from: json, torrentId: torrentId) else {
            throw DebridError.torrentNotFound(torrentId)
        }

        let state = (torrentObject["download_state"] as? String) ?? ""
        let files = parseFiles(from: torrentObject["files"])
        return TorrentSnapshot(state: state, files: files)
    }

    private func extractTorrentObject(from json: [String: Any], torrentId: String) -> [String: Any]? {
        if let object = json["data"] as? [String: Any] {
            return object
        }

        guard let list = json["data"] as? [[String: Any]] else {
            return nil
        }

        if let torrentID = Int64(torrentId),
           let match = list.first(where: { int64Value($0["id"]) == torrentID }) {
            return match
        }

        return list.first
    }

    private func parseFiles(from rawFiles: Any?) -> [TorrentFileEntry] {
        guard let rawFiles = rawFiles as? [[String: Any]] else {
            return []
        }

        return rawFiles.compactMap { file in
            guard let fileId64 = int64Value(file["id"]) else {
                return nil
            }

            let name = (file["name"] as? String) ?? (file["short_name"] as? String) ?? "Unknown"
            let size = int64Value(file["size"]) ?? 0

            return TorrentFileEntry(
                id: Int(fileId64),
                name: name,
                sizeBytes: size
            )
        }
    }

    private func bestFile(from files: [TorrentFileEntry]) -> TorrentFileEntry? {
        guard !files.isEmpty else { return nil }

        let candidates = files.map { file in
            DebridFileCandidate(
                link: String(file.id),
                fileName: file.name,
                sizeBytes: file.sizeBytes
            )
        }

        guard let selected = DebridFileSelector.selectBest(from: candidates),
              let selectedId = Int(selected.link) else {
            return nil
        }

        return files.first { $0.id == selectedId }
    }

    private func int64Value(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber {
            return number.int64Value
        }
        if let intValue = value as? Int {
            return Int64(intValue)
        }
        if let stringValue = value as? String {
            return Int64(stringValue)
        }
        return nil
    }
}
